packages/pds/src/label-cache.ts

```diff
@@ -0,0 +1,90 @@
+import { createDeferrable, wait } from '@atproto/common'
+import Database from './db'
+import { Label } from './db/tables/label'
+
+export class LabelCache {
+  bySubject: Record<string, Label[]> = {}
+  latestLabel = ''
+  defer = createDeferrable()
+  refreshes = 0
+
+  destroyed = false
+
+  constructor(public db: Database) {}
+
+  async start() {
+    await this.fullRefresh()
+    this.poll()
+  }
+
+  async fullRefresh() {
+    const allLabels = await this.db.db.selectFrom('label').selectAll().execute()
+    this.wipeCache()
+    this.processLabels(allLabels)
+  }
+
+  async partialRefresh() {
+    const labels = await this.db.db
+      .selectFrom('label')
+      .selectAll()
+      .where('cts', '>', this.latestLabel)
+      .execute()
+    this.processLabels(labels)
+  }
+
+  async poll() {
+    if (this.destroyed) return
+    if (this.refreshes >= 120) {
+      await this.fullRefresh()
+      this.refreshes = 0
+    } else {
+      await this.partialRefresh()
+      this.refreshes++
+    }
+    this.defer = createDeferrable()
+    await wait(500)
+    this.poll()
```

Comment:
Same down here, a `catch()` might be nice to avoid an unhandled rejection (admittedly unlikely).